### DISCUSSION TITLE
Credentials

### DISCUSSION THREAD

**Discussion by AbuSantos** [nextauthjs/next-auth#11812] [Help]
Hey guys, not sure whats wrong with my login, it works perfectly well on localhost, but when I deploy to vercel, it successfully retrieve the data but ,logs out almost immediately, I tried adding the NEXTAUTH_URL but still getting the same issue,

it works when i use localhost:3000 as the NEXTAUTH_URL but soon as i add my vercel.app domain, it fails to load the dashboard

import NextAuth from "next-auth";
import authConfig from "@/auth.config";
import { getToken } from "next-auth/jwt";
import {
  DEFAULT_LOGIN_REDIRECT,
  apiAuthPrefix,
  authRoutes,
  publicRoutes,
  DEFAULT_USER_ROLE,
} from "@/routes";
import { NextRequest } from "next/server";

const { auth } = NextAuth(authConfig);
//@ts-ignore

export default auth(async (req) => {
  const { nextUrl } = req;
  // console.log(req.url, "request url");
  // console.log(nextUrl, "next url");
  try {
    const token = await getToken({
      req,
      //@ts-ignore
      secret: process.env.NEXTAUTH_SECRET,
    });
    const userRole = token?.role;

    const isLoggedIn = !!token;

    const isAPIAuthRoute = nextUrl.pathname.startsWith(apiAuthPrefix);

    //   //if the nexturl.pathname is included in the publicroutes array, then it requires no auth
    const isPublicRoute = publicRoutes.includes(nextUrl.pathname);

    //   //if the nexturl.pathname is included in the authroutes array, then it requires auth
    // const isAuthRoute = authRoutes.includes(nextUrl.pathname);
    if (isAPIAuthRoute) {
      return null;
    }

    // Matching for both static and dynamic routes
    const isAuthRoute = authRoutes.some((route) =>
      nextUrl.pathname.startsWith(route)
    );

    if (isAuthRoute) {
      console.log(isAuthRoute, "auth route");
      if (isLoggedIn) {
        console.log(isLoggedIn, "login boolean");

        if (userRole === "parent") {
          console.log("Redirecting to parent page");
          return Response.redirect(new URL(DEFAULT_USER_ROLE, nextUrl));
        }

        return Response.redirect(new URL(DEFAULT_LOGIN_REDIRECT, nextUrl));
      }

      return null;
    }

    // console.log(userRole, "user role 2");

    if (!isLoggedIn && !isPublicRoute) {
      //taking users back to the previous used route
      let callbackUrl = nextUrl.pathname;
      if (nextUrl.search) {
        callbackUrl += nextUrl.search;
      }

      const encodeCallbackUrl = encodeURIComponent(callbackUrl);
      return Response.redirect(
        new URL(`/login?callbackUrl=${encodeCallbackUrl}`, nextUrl)
      );
    }

    return null;
  } catch (error) {
    console.error("Error fetching token:", error);
    // Handle error, e.g., redirect to login or display an error message
    return Response.redirect(new URL("/login", nextUrl));
  }
});

export const config = {
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
  unstable_allowDynamic: [
    "mongoose/dist/browser.umd.js",
    "./(models)/Parent.ts",
    "/node_modules/function-bind/**", // use a glob to allow anything in the function-bind 3rd party module
  ],
};
 this is my middleware. any help would be really appreciated

**Comment by Icegreeen** [comment]
See my comment on this issue here and see if this is your problem, if so, the solution I found will solve the problem:

https://github.com/nextauthjs/next-auth/discussions/7985#discussioncomment-10429041

@AbuSantos

**Reply by AbuSantos** [reply]
No, it doesn't, I  get no errors, it just stays on the login screen, when I change the NEXTAUTH_URL to localhost, it works well but when I change it back to myapp.vercel.app, it stays on the login screen.
